File: workers/ruby/lib/tasker_core/worker/event_poller.rb
# frozen_string_literal: true

module TaskerCore
  module Worker
    # EventPoller polls for step execution events from Rust via FFI
    #
    # Solves the cross-thread communication issue between Rust and Ruby by
    # actively polling the Rust worker for pending events. Events are forwarded
    # to EventBridge for normal processing.
    #
    # The poller runs in a dedicated background thread, continuously checking
    # for new events from the Rust worker at a configurable interval (default 10ms).
    # This approach allows Ruby to control the thread context and safely process
    # events without running into Ruby's Global Interpreter Lock (GIL) limitations.
    #
    # @example Starting the poller
    #   poller = TaskerCore::Worker::EventPoller.instance
    #   poller.start!
    #   # => Spawns polling thread with 10ms interval
    #
    #   # Check if poller started successfully
    #   if poller.active?
    #     puts "Event poller is running"
    #   end
    #
    # @example Checking poller status
    #   poller.active?
    #   # => true/false
    #
    #   if poller.active? && poller.polling_thread&.alive?
    #     puts "Poller thread is healthy"
    #   end
    #
    # @example Stopping the poller
    #   poller.stop!
    #   # => Gracefully stops polling thread (max 5 second wait)
    #
    #   # Poller will wait for current poll to complete
    #   # and then cleanly shutdown
    #
    # Threading Model:
    # - **Main Thread**: Ruby application, handler execution
    # - **Polling Thread**: Dedicated background thread for event polling
    # - **Rust Threads**: Rust worker runtime (separate from Ruby)
    #
    # The polling thread:
    # 1. Calls `TaskerCore::FFI.poll_step_events` to check for events
    # 2. If event found, forwards to EventBridge for processing
    # 3. If no event, sleeps for POLL_INTERVAL
    # 4. On error, logs and sleeps for 10x POLL_INTERVAL
    #
    # Why Polling is Necessary:
    #
    # Ruby's GIL prevents Rust from directly calling Ruby methods from Rust threads.
    # Direct FFI callbacks from Rust to Ruby would either:
    # - Cause segfaults (unsafe thread access)
    # - Block indefinitely (GIL contention)
    # - Require complex unsafe code
    #
    # Polling allows:
    # - Ruby to control thread context
    # - Safe event processing in Ruby thread
    # - Simple, reliable cross-language communication
    # - Predictable performance characteristics
    #
    # Performance Characteristics:
    # - **Poll Interval**: 10ms (0.01 seconds) - configurable via POLL_INTERVAL
    # - **Max Latency**: ~10ms from event generation to processing start
    # - **CPU Usage**: Minimal (yields during sleep)
    # - **Error Recovery**: Automatic with exponential backoff (10x interval)
    #
    # Event Flow:
    # 1. Rust worker detects step ready for execution
    # 2. Rust queues event in internal event queue
    # 3. EventPoller calls poll_step_events via FFI
    # 4. Rust returns next event (or nil if queue empty)
    # 5. EventPoller forwards event to EventBridge
    # 6. EventBridge publishes to Ruby subscribers
    # 7. StepExecutionSubscriber executes handler
    #
    # @see TaskerCore::Worker::EventBridge For event processing
    # @see TaskerCore::FFI For Rust FFI operations
    # @see POLL_INTERVAL For polling frequency configuration
    class EventPoller
      include Singleton

      attr_reader :logger, :active, :polling_thread

      # Polling interval in seconds (10ms)
      # Lower values reduce latency but increase CPU usage
      # Higher values reduce CPU usage but increase latency
      POLL_INTERVAL = 0.01

      def initialize
        @logger = TaskerCore::Logger.instance
        @active = false
        @polling_thread = nil
      end

      # Start polling for events from Rust
      def start!
        return if @active

        @active = true
        logger.info 'Starting EventPoller - polling for step execution events from Rust'

        @polling_thread = Thread.new do
          poll_events_loop
        end

        logger.info '✅ EventPoller started successfully'
      end

      # Stop polling
      def stop!
        return unless @active

        logger.info 'Stopping EventPoller...'
        @active = false

        if @polling_thread&.alive?
          @polling_thread.join(5.0) # Wait up to 5 seconds for thread to finish
          @polling_thread.kill if @polling_thread.alive? # Force kill if still running
        end

        @polling_thread = nil
        logger.info '✅ EventPoller stopped'
      end

      # Check if poller is active
      def active?
        @active
      end

      private

      # Main polling loop - runs in dedicated thread
      def poll_events_loop
        logger.debug 'EventPoller: Starting poll loop'

        while @active
          begin
            # Poll for next event from Rust via FFI
            event_data = TaskerCore::FFI.poll_step_events

            if event_data.nil?
              # No events available, sleep briefly
              sleep(POLL_INTERVAL)
              next
            end

            # Process the event through EventBridge
            process_event(event_data)
          rescue StandardError => e
            logger.error "EventPoller error: #{e.message}"
            logger.error e.backtrace.join("\n")

            # Sleep longer on error to avoid tight error loops
            sleep(POLL_INTERVAL * 10) if @active
          end
        end

        logger.debug 'EventPoller: Poll loop terminated'
      end

      # Process a polled event through the EventBridge
      def process_event(event_data)
        logger.debug "EventPoller: Processing event from Rust: #{event_data}"

        # Forward to EventBridge for normal processing
        EventBridge.instance.publish_step_execution(event_data)

        logger.debug 'EventPoller: Event forwarded to EventBridge successfully'
      rescue StandardError => e
        logger.error "EventPoller: Failed to process event: #{e.message}"
        logger.error e.backtrace.join("\n")
        raise
      end
    end
  end
end

File: workers/python/python/tasker_core/event_poller.py
"""Threaded event poller for step execution events.

This module provides the EventPoller class that polls for step events
from the FFI dispatch channel and emits them to registered handlers.

Example:
    >>> from tasker_core import EventPoller
    >>>
    >>> def on_step_event(event):
    ...     print(f"Processing step {event.step_uuid}")
    ...
    >>> poller = EventPoller()
    >>> poller.on_step_event(on_step_event)
    >>> poller.start()
    >>> # ... events flow to handlers ...
    >>> poller.stop()
"""

from __future__ import annotations

import threading
import time
from collections.abc import Callable
from typing import TYPE_CHECKING, Any

from tasker_core._tasker_core import (
    check_starvation_warnings as _check_starvation_warnings,
)
from tasker_core._tasker_core import (
    cleanup_timeouts as _cleanup_timeouts,
)
from tasker_core._tasker_core import (
    get_ffi_dispatch_metrics as _get_ffi_dispatch_metrics,
)
from tasker_core._tasker_core import (
    poll_step_events as _poll_step_events,
)

from .logging import log_debug, log_error, log_info, log_warn
from .types import FfiDispatchMetrics, FfiStepEvent

if TYPE_CHECKING:
    pass


# Type aliases for callbacks
StepEventCallback = Callable[[FfiStepEvent], None]
MetricsCallback = Callable[[FfiDispatchMetrics], None]
ErrorCallback = Callable[[Exception], None]


class EventPoller:
    """Threaded event poller for step execution events.

    The EventPoller runs in a separate thread and polls for step events
    from the FFI dispatch channel. Events are emitted to registered
    handlers via callbacks.

    This matches the Ruby EventPoller pattern with a dedicated polling
    thread that yields control between polls.

    Threading Model:
        - Main Thread: Python application, handler execution
        - Polling Thread: Dedicated background thread for event polling
        - Rust Threads: Rust worker runtime (separate from Python)

    Why Polling is Necessary:
        Python's GIL prevents Rust from directly calling Python methods
        from Rust threads. Polling allows Python to control thread context
        and safely process events.

    Performance Characteristics:
        - Poll Interval: 10ms (configurable)
        - Max Latency: ~10ms from event generation to processing start
        - CPU Usage: Minimal (yields during sleep)
        - Error Recovery: Automatic with exponential backoff

    Example:
        >>> poller = EventPoller(polling_interval_ms=10)
        >>> poller.on_step_event(handle_step)
        >>> poller.on_error(handle_error)
        >>> poller.start()
        >>>
        >>> # Events flow to handle_step callback
        >>>
        >>> poller.stop()

    Attributes:
        polling_interval_ms: Milliseconds between polls when no events.
        starvation_check_interval: Polls between starvation checks.
        cleanup_interval: Polls between timeout cleanups.
    """

    # Default polling interval in milliseconds
    DEFAULT_POLL_INTERVAL_MS = 10

    # Check for starvation warnings every N poll iterations
    # At 10ms poll interval, 100 iterations = 1 second
    DEFAULT_STARVATION_CHECK_INTERVAL = 100

    # Cleanup timed-out events every N poll iterations
    # At 10ms poll interval, 1000 iterations = 10 seconds
    DEFAULT_CLEANUP_INTERVAL = 1000

    def __init__(
        self,
        polling_interval_ms: int = DEFAULT_POLL_INTERVAL_MS,
        starvation_check_interval: int = DEFAULT_STARVATION_CHECK_INTERVAL,
        cleanup_interval: int = DEFAULT_CLEANUP_INTERVAL,
    ) -> None:
        """Initialize the EventPoller.

        Args:
            polling_interval_ms: Milliseconds between polls when no events.
            starvation_check_interval: Number of polls between starvation checks.
            cleanup_interval: Number of polls between timeout cleanups.
        """
        self._polling_interval = polling_interval_ms / 1000.0
        self._starvation_check_interval = starvation_check_interval
        self._cleanup_interval = cleanup_interval
        self._running = False
        self._thread: threading.Thread | None = None
        self._poll_count = 0

        # Callbacks
        self._step_event_callbacks: list[StepEventCallback] = []
        self._metrics_callbacks: list[MetricsCallback] = []
        self._error_callbacks: list[ErrorCallback] = []

    def on_step_event(self, callback: StepEventCallback) -> None:
        """Register a callback for step events.

        The callback will be invoked on the polling thread when a step
        event is received from the FFI dispatch channel.

        Args:
            callback: Function that accepts an FfiStepEvent.

        Example:
            >>> def handle_step(event: FfiStepEvent):
            ...     print(f"Processing {event.step_uuid}")
            >>> poller.on_step_event(handle_step)
        """
        self._step_event_callbacks.append(callback)

    def on_metrics(self, callback: MetricsCallback) -> None:
        """Register a callback for metrics updates.

        The callback is invoked periodically with FFI dispatch metrics.

        Args:
            callback: Function that accepts FfiDispatchMetrics.

        Example:
            >>> def handle_metrics(metrics: FfiDispatchMetrics):
            ...     if metrics.starvation_detected:
            ...         print(f"Warning: {metrics.starving_event_count} starving")
            >>> poller.on_metrics(handle_metrics)
        """
        self._metrics_callbacks.append(callback)

    def on_error(self, callback: ErrorCallback) -> None:
        """Register a callback for errors.

        The callback is invoked when an error occurs during polling.

        Args:
            callback: Function that accepts an Exception.

        Example:
            >>> def handle_error(error: Exception):
            ...     log_error(f"Poller error: {error}")
            >>> poller.on_error(handle_error)
        """
        self._error_callbacks.append(callback)

    def start(self) -> None:
        """Start the event polling thread.

        Creates a daemon thread that continuously polls for events.
        The thread will stop when stop() is called or the main program exits.

        Raises:
            RuntimeError: If the poller is already running.

        Example:
            >>> poller.start()
            >>> assert poller.is_running
        """
        if self._running:
            raise RuntimeError("EventPoller already running")

        self._running = True
        self._thread = threading.Thread(
            target=self._poll_loop,
            name="tasker-event-poller",
            daemon=True,
        )
        self._thread.start()

        log_info("EventPoller started", {"interval_ms": str(int(self._polling_interval * 1000))})

    def stop(self, timeout: float = 5.0) -> None:
        """Stop the event polling thread.

        Signals the polling thread to stop and waits for it to finish.
        Safe to call even if the poller is not running.

        Args:
            timeout: Maximum seconds to wait for thread to stop.

        Example:
            >>> poller.stop()
            >>> assert not poller.is_running
        """
        if not self._running:
            return

        log_info("Stopping EventPoller...")
        self._running = False

        if self._thread is not None:
            self._thread.join(timeout=timeout)
            self._thread = None

        log_info("EventPoller stopped")

    @property
    def is_running(self) -> bool:
        """Check if the poller is currently running.

        Returns:
            True if the polling thread is active.
        """
        return (
            self._running
            and self._thread is not None
            and self._thread.is_alive()
        )

    def get_metrics(self) -> FfiDispatchMetrics | None:
        """Get current FFI dispatch metrics.

        This can be called from any thread to check the current state
        of the dispatch channel.

        Returns:
            FfiDispatchMetrics if available, None on error.

        Example:
            >>> metrics = poller.get_metrics()
            >>> if metrics and metrics.starvation_detected:
            ...     print("Starvation detected!")
        """
        try:
            metrics_dict = _get_ffi_dispatch_metrics()
            return FfiDispatchMetrics.model_validate(metrics_dict)
        except Exception as e:
            log_debug(f"Failed to get FFI dispatch metrics: {e}")
            return None

    def _poll_loop(self) -> None:
        """Main polling loop (runs in separate thread)."""
        log_debug("EventPoller: Starting poll loop")

        while self._running:
            try:
                self._poll_count += 1

                # Poll for next event
                event_data = _poll_step_events()

                if event_data is not None:
                    self._process_event(event_data)
                else:
                    # No event available, sleep before next poll
                    time.sleep(self._polling_interval)

                # Periodic starvation check
                if self._poll_count % self._starvation_check_interval == 0:
                    self._check_starvation()

                # Periodic cleanup of timed-out events
                if self._poll_count % self._cleanup_interval == 0:
                    self._cleanup_timeouts()

            except Exception as e:
                self._emit_error(e)
                # Sleep longer on error to avoid tight error loops
                time.sleep(self._polling_interval * 10)

        log_debug("EventPoller: Poll loop terminated")

    def _process_event(self, event_data: dict[str, Any]) -> None:
        """Process a polled event through callbacks."""
        try:
            event = FfiStepEvent.model_validate(event_data)
            log_debug(
                "EventPoller: Processing event",
                {"event_id": event.event_id, "step_uuid": event.step_uuid},
            )

            for callback in self._step_event_callbacks:
                try:
                    callback(event)
                except Exception as e:
                    log_error(
                        f"EventPoller: Callback error: {e}",
                        {"event_id": event.event_id},
                    )
                    self._emit_error(e)

        except Exception as e:
            log_error(f"EventPoller: Failed to process event: {e}")
            self._emit_error(e)

    def _check_starvation(self) -> None:
        """Check for starvation warnings and emit metrics."""
        try:
            # Trigger Rust-side starvation logging
            _check_starvation_warnings()

            # Get and emit metrics
            metrics_dict = _get_ffi_dispatch_metrics()
            metrics = FfiDispatchMetrics.model_validate(metrics_dict)

            for callback in self._metrics_callbacks:
                try:
                    callback(metrics)
                except Exception as e:
                    log_debug(f"EventPoller: Metrics callback error: {e}")

            if metrics.starvation_detected:
                log_warn(
                    "EventPoller: Starvation detected",
                    {
                        "pending_count": str(metrics.pending_count),
                        "starving_count": str(metrics.starving_event_count),
                    },
                )

        except Exception as e:
            # Don't let starvation check errors affect polling
            log_debug(f"EventPoller: Starvation check error (non-fatal): {e}")

    def _cleanup_timeouts(self) -> None:
        """Clean up timed-out pending events."""
        try:
            count = _cleanup_timeouts()
            if count > 0:
                log_warn(
                    "EventPoller: Cleaned up timed-out events",
                    {"count": str(count)},
                )
        except Exception as e:
            log_debug(f"EventPoller: Cleanup error (non-fatal): {e}")

    def _emit_error(self, error: Exception) -> None:
        """Emit an error to registered callbacks."""
        import contextlib

        for callback in self._error_callbacks:
            # Use contextlib.suppress to ignore errors in error callbacks
            # (we don't want error callbacks to cause more errors)
            with contextlib.suppress(Exception):
                callback(error)


__all__ = ["EventPoller"]
